### PR TITLE
Add User Input block to Admin Settings, use query param for state, and add redirect URL.

### PR DESCRIPTION
**Pull request by kostyalmm**
## Summary

Addresses issue #2041

## Relevant technical choices

* Add `UserInputSettings` notification to Admin Settings when the user input is incomplete.
* Add `UserInputPreview` to Admin Setting when the user input is completed.
* Hide `UserInputPreview` footer in Admin Settings.
* Add `useQueryString` hook to use the query param as state
* Add a `question` query param state to `UserInputQuestionnaire`
* Add a `redirect_url` query param to `UserInputQuestionnaire`
* Go to `${ userInputURL }&question=${ question }` when clicked on edit button in `AdminSettings`.
* After clicking submit, go to the Admin Settings using the `redirect_url` query param.

Reason why `pbeshai/use-query-params` wasn't used: We need to wrap the app in a Context for this to work. `react-use` doesn't have the ability to use query param as state and using hash would have resulted in problem with methods like `addQueryArgs` (it converts # to %23).

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/sass/components/settings/_googlesitekit-settings-user-input.scss [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+.googlesitekit-settings-user-input {
+
+	.googlesitekit-user-input__preview {
+		padding: 25px !important;
+	}
+}
```
Comment (eugene-manuilov) [inline assets/sass/components/settings/_googlesitekit-settings-user-input.scss:24]: Just start your rule from `.googlesitekit-plugin` and you won't need `!important`:

```suggestion
.googlesitekit-plugin {
    .googlesitekit-settings-user-input {
        .googlesitekit-user-input__preview {
            padding: 25px;
        }
    }
}
```

File: assets/js/hooks/useQueryString.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		},
+		[ key ]
+	);
+
+	return [ value, onSetValue ];
+}
+
+const setQueryStringWithoutPageReload = ( qsValue ) => {
+	const newURL = `${ global.location.pathname }${ qsValue }`;
+	global.history.pushState( { path: newURL }, '', newURL );
```
Comment (eugene-manuilov) [inline assets/js/hooks/useQueryString.js:53]: @kostyalmm use `replaceState` instead of `pushState` to eliminate browser history pollution.

File: assets/js/components/settings/settings-admin.js [hunk 3]
```diff
@@ -19,20 +19,68 @@
...
+import { STORE_NAME as CORE_USER } from '../../googlesitekit/datastore/user/constants';
 import Layout from '../layout/layout';
 import OptIn from '../optin';
 import VisuallyHidden from '../VisuallyHidden';
 import ResetButton from '../ResetButton';
+import UserInputPreview from '../user-input/UserInputPreview';
+import {
+	USER_INPUT_QUESTION_GOALS, USER_INPUT_QUESTION_HELP_NEEDED,
+	USER_INPUT_QUESTION_POST_FREQUENCY,
+	USER_INPUT_QUESTION_ROLE, USER_INPUT_QUESTION_SEARCH_TERMS,
```
Comment (eugene-manuilov) [inline assets/js/components/settings/settings-admin.js:38]: Please, use one import per line if you expand named imports into multiple lines.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @kostyalmm! I have added some comments that need to be addressed before I can approve this PR. Could you please look at it?

File: assets/js/components/user-input/UserInputQuestionnaire.js [hunk 4]
```diff
@@ -38,17 +39,43 @@ import {
...
+				return i;
+			}
+		}
+	} );
+
+	useEffect( () => {
+		if ( activeSlugIndex > answeredUntilIndex ) {
+			setActiveSlug( steps[ answeredUntilIndex ] );
+		}
+	}, [ answeredUntilIndex ] );
```
[the remark this answers is not in the document]
Comment (kostyalmm) [reply assets/js/components/user-input/UserInputQuestionnaire.js:78]: So if someone tries accessing `question=5` through the URL and questions 1-4 are not filled out, this logic will put them on `question=1`.

Also if someone is on question 5, and then hits reload, we want them to re fill the form from state as it is stored in the local state.

Unless and until, questions until `n` are filled, the user should not be able to see/fill `n+1`th question.

[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/components/user-input/UserInputQuestionnaire.js:78]: Ok, thanks!

File: assets/js/components/settings/settings-admin.js [hunk 5]
```diff
@@ -19,20 +19,70 @@
...
+import UserInputPreview from '../user-input/UserInputPreview';
+import {
+	USER_INPUT_QUESTION_GOALS,
+	USER_INPUT_QUESTION_HELP_NEEDED,
+	USER_INPUT_QUESTION_POST_FREQUENCY,
+	USER_INPUT_QUESTION_ROLE,
+	USER_INPUT_QUESTION_SEARCH_TERMS,
+} from '../user-input/util/constants';
+import { STORE_NAME as CORE_SITE } from '../../googlesitekit/datastore/site/constants';
+import { addQueryArgs } from '@wordpress/url';
```
Comment (eugene-manuilov) [inline assets/js/components/settings/settings-admin.js:44]: @kostyalmm, please, move it to the WordPress dependencies section.

File: assets/js/components/settings/settings-admin.js [hunk 6]
```diff
@@ -19,20 +19,70 @@
...
+	USER_INPUT_QUESTION_SEARCH_TERMS,
+];
 
 const SettingsAdmin = () => {
+	const { isUserInputCompleted, userInputURL } = useSelect( ( select ) => {
+		return {
+			isUserInputCompleted: select( CORE_USER ).getUserInputState() === 'completed',
+			userInputURL: select( CORE_SITE ).getAdminURL( 'googlesitekit-user-input' ),
+		};
+	} );
```
Comment (eugene-manuilov) [inline assets/js/components/settings/settings-admin.js:62]:
```suggestion
    const { isUserInputCompleted, userInputURL } = useSelect( ( select ) => ( {
        isUserInputCompleted: select( CORE_USER ).getUserInputState() === 'completed',
        userInputURL: select( CORE_SITE ).getAdminURL( 'googlesitekit-user-input' ),
    } ) );
```

File: assets/js/hooks/useQueryString.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+			setValue( newValue );
+			setQueryStringValue( key, newValue );
+		},
+		[ key ]
+	);
+
+	return [ value, onSetValue ];
+}
+
+const setQueryStringWithoutPageReload = ( qsValue ) => {
```
Comment (eugene-manuilov) [inline assets/js/hooks/useQueryString.js:46]: @kostyalmm, could you please convert `setQueryStringWithoutPageReload`, `getQueryStringValue` and `setQueryStringValue` functions to be regular functions (not arrow ones) and add jsdoc to it.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Awesome work, @kostyalmm! Just added a few stylistic requests and I will approve it.

File: assets/js/hooks/useQueryString.js (hunk 7 above)
Comment (kostyalmm) [reply assets/js/hooks/useQueryString.js]: I actually ended up removing those methods and putting them inside the hook with `useCallback`. Do you want me to add the jsDoc even now? or can we skip that?

File: orders_export.csv [hunk 8]
```diff
@@ -0,0 +1,11 @@
+Name,Email,Financial Status,Paid at,Fulfillment Status,Fulfilled at,Accepts Marketing,Currency,Subtotal,Shipping,Taxes,Total,Discount Code,Discount Amount,Shipping Method,Created at,Lineitem quantity,Lineitem name,Lineitem price,Lineitem compare at price,Lineitem sku,Lineitem requires shipping,Lineitem taxable,Lineitem fulfillment status,Billing Name,Billing Street,Billing Address1,Billing Address2,Billing Company,Billing City,Billing Zip,Billing Province,Billing Country,Billing Phone,Shipping Name,Shipping Street,Shipping Address1,Shipping Address2,Shipping Company,Shipping City,Shipping Zip,Shipping Province,Shipping Country,Shipping Phone,Notes,Note Attributes,Cancelled at,Payment Method,Payment Reference,Refunded Amount,Vendor,Outstanding Balance,Employee,Location,Device ID,Id,Tags,Risk Level,Source,Lineitem discount,Tax 1 Name,Tax 1 Value,Tax 2 Name,Tax 2 Value,Tax 3 Name,Tax 3 Value,Tax 4 Name,Tax 4 Value,Tax 5 Name,Tax 5 Value,Phone,Receipt Number
```
Comment (eugene-manuilov) [inline orders_export.csv:1]: Do we need this file? Probably it has been committed by accident, hasn't it?

File: assets/js/hooks/useQueryString.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * @param {string}        key          The query param key to be used.
+ * @param {(string|null)} initialValue The initial value for the query param to be used.
+ * @return {Array} The getter and setter for the query param state.
+ */
+function useQueryString( key, initialValue = null ) {
+	const getQueryStringValue = useCallback( () => {
+		const urlParams = new URLSearchParams( global.location.search );
+		return urlParams.get( key );
+	}, [ global.location.search ] );
```
Comment (eugene-manuilov) [inline assets/js/hooks/useQueryString.js:37]: Looks like we don't need this function and can rework `useState` initial value to be something like this:

```js
const [ value, setValue ] = useState(
    new URLSearchParams( global.location.search ).get( key ) || initialValue,
);

```

File: assets/js/hooks/useQueryString.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+	const setQueryStringValue = useCallback( ( value ) => {
+		const urlParams = new URLSearchParams( global.location.search );
+		urlParams.set( key, value );
+		const newURL = `${ global.location.pathname }?${ urlParams.toString() }`;
+		global.history.replaceState( { path: newURL }, '', newURL );
+	}, [ global.location.pathname, global.location.search ] );
```
Comment (eugene-manuilov) [inline assets/js/hooks/useQueryString.js:44]: Same here, I don't think we need this callback function. We can incorporate it into the `onSetValue`:

```js
    const onSetValue = useCallback( ( newValue ) => {
        setValue( newValue );

        const urlParams = new URLSearchParams( global.location.search );
        urlParams.set( key, newValue );

        const path = `${ global.location.pathname }?${ urlParams.toString() }`;
        global.history.replaceState( { path }, '', path );
}, [ key ] );
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]

File: assets/js/hooks/useQueryString.js (hunk 7 above)
Comment (eugene-manuilov) [reply assets/js/hooks/useQueryString.js]: No, we don't need jsdoc then. :+1:

File: assets/js/hooks/useQueryString.js (hunk 9 above)
Comment (kostyalmm) [reply assets/js/hooks/useQueryString.js]: Made the changes. Also I have deleted your previous comment 😅, it was a sensitive file so didn't want it to have a reference. 😅

**Review by eugene-manuilov — APPROVED** [review]
Awesome! Thanks!

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
@kostyalmm! Could you please fix broken js tests?

**Review by eugene-manuilov — DISMISSED** [review]
Thanks, @kostyalmm!

File: assets/js/hooks/useQueryString.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+ * Uses a location query param as a state.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string}        key          The query param key to be used.
+ * @param {(string|null)} initialValue The initial value for the query param to be used.
+ * @return {Array} The getter and setter for the query param state.
+ */
+function useQueryString( key, initialValue = null ) {
+	const [ value, setValue ] = useState( new URLSearchParams( global.location.search ).get( key ) || initialValue );
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryString.js:34]: Please use [`getQueryArg`](https://github.com/WordPress/gutenberg/tree/master/packages/url#getQueryArg) instead.

File: assets/js/hooks/useQueryString.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+ */
+import { useState, useCallback } from '@wordpress/element';
+
+/**
+ * Uses a location query param as a state.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string}        key          The query param key to be used.
+ * @param {(string|null)} initialValue The initial value for the query param to be used.
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryString.js:30]: This should be marked as an optional parameter.
```suggestion
 * @param {string} [initialValue] Optional. The initial value for the query param to be used. Default: `null`.
```

File: assets/js/hooks/useQueryString.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ * @return {Array} The getter and setter for the query param state.
+ */
+function useQueryString( key, initialValue = null ) {
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryString.js:33]: Can we name this `useQueryArg` instead for consistency with the `(add|get)QueryArg(s)` functions that we already use in many places? I think that would be a slightly more accurate name as well since we're getting the value of a single query arg rather than the entire query string.

File: assets/js/hooks/useQueryString.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	const [ value, setValue ] = useState( new URLSearchParams( global.location.search ).get( key ) || initialValue );
+	const onSetValue = useCallback(
+		( newValue ) => {
+			setValue( newValue );
+
+			const urlParams = new URLSearchParams( global.location.search );
+			urlParams.set( key, newValue );
+
+			const newURL = `${ global.location.pathname }?${ urlParams.toString() }`;
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryString.js:42]: Please use [`addQueryArgs`](https://github.com/WordPress/gutenberg/tree/master/packages/url#addQueryArgs) instead: e.g.
```js
const newURL = addQueryArgs( global.location.href, { [ key ]: newValue } )
```

File: assets/js/hooks/useQueryString.test.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * Internal dependencies
+ */
+import { renderHook, actHook } from '../../../tests/js/test-utils';
+import useQueryString from './useQueryString';
+
+describe( 'useQueryString', () => {
+	let oldLocation;
+	let oldHistory;
+	const historyPushMock = jest.fn();
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryString.test.js:28]: Let's name this to match the method more closely. Otherwise it might be confused as a mock for `history.pushState`.

```suggestion
	const historyReplaceStateMock = jest.fn();
```

File: assets/js/hooks/useQueryString.test.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+		actHook( () => {
+			setQuery( 'prod' );
+		} );
+
+		const updatedURL = 'path?page=prod';
+		expect( historyPushMock ).toHaveBeenCalledWith( {
+			path: updatedURL,
+		}, '', updatedURL );
+	} );
+} );
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryString.test.js:69]: Let's add another test here to make sure that `setQuery` URL-encodes the value (done automatically by `addQueryArgs`).

File: assets/sass/components/settings/_googlesitekit-settings-user-input.scss [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+.googlesitekit-plugin {
+
+	.googlesitekit-settings-user-input {
+
+		.googlesitekit-user-input__preview {
+			padding: 25px !important;
```
Comment (aaemnnosttv) [inline assets/sass/components/settings/_googlesitekit-settings-user-input.scss:24]: Is there not a scss variable we can use for the value instead? It looks close to `$grid-gap-desktop` – if so, it may also need to use `$grid-gap-phone` in smaller viewports.

Also, why `!important` ?

File: assets/js/components/settings/settings-admin.js [hunk 18]
```diff
@@ -19,20 +19,68 @@
...
+import UserInputSettings from '../notifications/UserInputSettings';
+const { useSelect } = Data;
+
+const questions = [
+	USER_INPUT_QUESTION_ROLE,
+	USER_INPUT_QUESTION_POST_FREQUENCY,
+	USER_INPUT_QUESTION_GOALS,
+	USER_INPUT_QUESTION_HELP_NEEDED,
+	USER_INPUT_QUESTION_SEARCH_TERMS,
+];
```
Comment (aaemnnosttv) [inline assets/js/components/settings/settings-admin.js:54]: This array is duplicated with the same array in `UserInputQuestionnaire` https://github.com/google/site-kit-wp/blob/878a304b92fd13a642894c6938657e1e1932c5f9/assets/js/components/user-input/UserInputQuestionnaire.js#L47-L53

This risks these two being out of sync if the array changes in one place and not the other.

Let's extract this array into the user input constants and reference that in both places rather than duplicating it in both.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @kostyalmm – this is off to a good start, just a few points to address.

File: assets/js/components/user-input/UserInputPreview.js [hunk 19]
```diff
@@ -58,7 +58,12 @@ export default function UserInputPreview( { back, goTo } ) {
 		setIsNavigating( true );
 		const response = await saveUserInputSettings();
 		if ( ! response.error ) {
-			global.location.assign( addQueryArgs( dashboardURL, { notification: 'user_input_success' } ) );
+			let [ sanitizedRedirectURL, hash ] = ( redirectURL ?? dashboardURL ).split( '#' );
+			sanitizedRedirectURL = addQueryArgs( sanitizedRedirectURL, { notification: 'user_input_success' } );
+			if ( hash ) {
+				sanitizedRedirectURL += `#${ hash }`;
+			}
```
[the remark this answers is not in the document]
Comment (kostyalmm) [reply assets/js/components/user-input/UserInputPreview.js:65]: As you can see [here](https://github.com/WordPress/gutenberg/blob/master/packages/url/src/add-query-args.js#L23), the `addQueryArgs` does not care about the `hash` at all. 
It takes the url, appends query arg to the url. 
What happens in the case of hash is this:
if 
```js
url = http://example.com/path?argument=value#hash
```
then 
```js
addQueryArgs(url, {test: "test"}) = "http://example.com/path?argument=value%23hash&test=test"
```

The `#` gets sandwiched into a query arg itself. What do you suggest I do here? Splitting seems to be the best thing as we need to get rid of the hash before adding query args to it and then add the hash later.

File: assets/js/components/user-input/UserInputQuestionnaire.js [hunk 20]
```diff
@@ -38,17 +39,43 @@ import {
...
 		USER_INPUT_QUESTION_SEARCH_TERMS,
 	];
+	const steps = [ ...questions, 'preview' ];
+
+	const [ activeSlug, setActiveSlug ] = useQueryString( 'question', steps[ 0 ] );
+	const [ redirectURL ] = useQueryString( 'redirect_url' );
+
+	const activeSlugIndex = steps.indexOf( activeSlug );
+
+	if ( activeSlugIndex === -1 ) {
```
[the remark this answers is not in the document]
Comment (kostyalmm) [reply assets/js/components/user-input/UserInputQuestionnaire.js:61]: We need the index later in the component, that is why it is done like this.

File: assets/sass/components/settings/_googlesitekit-settings-user-input.scss (hunk 17 above)
Comment (kostyalmm) [reply assets/sass/components/settings/_googlesitekit-settings-user-input.scss]: The `googlesitekit-user-input__preview` class has a hardcoded value of `padding: 50px 23px;` that is why I used `25px`. I used the variables instead, please check.

also removed the `!important`, it was not required 😅

File: assets/js/components/user-input/UserInputPreview.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/components/user-input/UserInputPreview.js:65]: I see what you mean @kostyalmm . It works fine when the hash is before the query string but is not handled when it is after which seems to be the correct place when both are present 🤔 

This seems to be a known bug in the package: https://github.com/WordPress/gutenberg/issues/16655

I feel like the safest way is to use `URL` rather than assume the hash is in a particular spot. This seems to be the direction the PR for that issue is taking as well.

For this specific case, I think we can avoid using `addQueryArgs` since we're simply adding a specific, known parameter which also doesn't require encoding.

```js
const url = new URL( redirectURL || dashboardURL );
// Here we don't use `addQueryArgs` due to a bug with how it handles hashes
// See https://github.com/WordPress/gutenberg/issues/16655
url.searchParams.set( 'notification', 'user_input_success' );
const finalURL = url.toString();
```

File: assets/js/hooks/useQueryArg.js [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+import { useState, useCallback } from '@wordpress/element';
+import { addQueryArgs, getQueryArg } from '@wordpress/url';
+
+/**
+ * Uses a location query param as a state.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} key            The query param key to be used.
+ * @param {string} [initialValue] Optional. The initial value for the query param to be used. Default: `null`.
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryArg.js:31]: On second thought, I think `undefined` is correct for the initial value here since a `null` value will result in the query variable being added with no value – e.g. `?foo=`, where `undefined` is more accurate to the key not existing.

```suggestion
 * @param {string} [initialValue] Optional. The initial value for the query param to be used.
```

File: assets/js/hooks/useQueryArg.js [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+ */
+function useQueryArg( key, initialValue ) {
+	const [ value, setValue ] = useState( getQueryArg( global.location.href, key ) || initialValue );
+
+	const onSetValue = useCallback(
+		( newValue ) => {
+			setValue( newValue );
+
+			const newURL = addQueryArgs( global.location.href, { [ key ]: newValue } );
+			global.history.replaceState( { path: newURL }, '', newURL );
```
Comment (aaemnnosttv) [inline assets/js/hooks/useQueryArg.js:42]: I'm not sure the first argument (`stateObj`) is needed here. Using `path` for a full URL isn't really accurate either.

Let's pass this as `null` for now since there isn't anything we currently need this for.
```suggestion
			global.history.replaceState( null, '', newURL );
```

File: assets/sass/components/settings/_googlesitekit-settings-user-input.scss [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+
+	.googlesitekit-settings-user-input {
+		padding-bottom: 0;
+
+		.googlesitekit-user-input__preview {
+			padding: $grid-gap-desktop;
+
+			@media (max-width: $bp-mobileOnly) {
+				padding: $grid-gap-phone;
+			}
```
Comment (aaemnnosttv) [inline assets/sass/components/settings/_googlesitekit-settings-user-input.scss:29]: This should be the other way around – mobile styles first and then media query with `min-width` for larger viewports 🙂  Almost all of our styles are this way.

File: assets/js/components/user-input/util/constants.js [hunk 24]
```diff
@@ -27,6 +27,14 @@ export const USER_INPUT_QUESTION_GOALS = 'goals';
 export const USER_INPUT_QUESTION_HELP_NEEDED = 'helpNeeded';
 export const USER_INPUT_QUESTION_SEARCH_TERMS = 'searchTerms';
 
+export const QUESTIONS = [
```
Comment (aaemnnosttv) [inline assets/js/components/user-input/util/constants.js:30]: Let's name this a little more specifically, for consistency with the others but also to make it a bit more clear what it is.

```suggestion
export const USER_INPUT_QUESTIONS_LIST = [
```

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @kostyalmm – a few additional comments to address but this is looking good!

File: assets/js/components/user-input/UserInputQuestionnaire.js [hunk 25]
```diff
@@ -19,36 +19,56 @@
...
+	} );
+
+	useEffect( () => {
+		if ( answeredUntilIndex === -1 ) {
+			return;
+		}
+		if ( activeSlugIndex > answeredUntilIndex ) {
+			setActiveSlug( steps[ answeredUntilIndex ] );
+		}
+	}, [ answeredUntilIndex ] );
```
[the remark this answers is not in the document]
Comment (kostyalmm) [reply assets/js/components/user-input/UserInputQuestionnaire.js:71]: I did not add this as I thought that the answers are set in the local state which will make the `answeredUntilIndex` a constant for the lifetime of the component. But as I saw, `getUserInputSettings` returns an object which has the questions answered in realtime therefore added `activeSlugIndex`. Thanks for the catch 👍

File: assets/js/hooks/useQueryArg.js [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { useState, useCallback } from '@wordpress/element';
+import { addQueryArgs, getQueryArg } from '@wordpress/url';
+
+/**
+ * Uses a location query param as a state.
```
Comment (tofumatt) [inline assets/js/hooks/useQueryArg.js:26]: I'm not sure what "as a state" here means. How about:

```suggestion
 * Uses a location query param as a variable in a component.
```

File: assets/js/hooks/useQueryArg.js [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+ * @since n.e.x.t
+ *
+ * @param {string} key            The query param key to be used.
+ * @param {string} [initialValue] Optional. The initial value for the query param to be used.
+ * @return {Array} The getter and setter for the query param state.
+ */
+function useQueryArg( key, initialValue ) {
+	const [ value, setValue ] = useState( getQueryArg( global.location.href, key ) || initialValue );
+
+	const onSetValue = useCallback(
```
Comment (tofumatt) [inline assets/js/hooks/useQueryArg.js:37]: This shouldn't need `useCallback`, so let's remove it.

File: assets/js/hooks/useQueryArg.js [hunk 28]
```diff
@@ -0,0 +1,50 @@
...
+		( newValue ) => {
+			setValue( newValue );
+
+			const newURL = addQueryArgs( global.location.href, { [ key ]: newValue } );
+			global.history.replaceState( null, '', newURL );
```
Comment (tofumatt) [inline assets/js/hooks/useQueryArg.js:42]: Nice 👍🏻

File: assets/js/hooks/useQueryArg.js [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Uses a location query param as a state.
+ *
+ * @since n.e.x.t
+ *
+ * @param {string} key            The query param key to be used.
+ * @param {string} [initialValue] Optional. The initial value for the query param to be used.
+ * @return {Array} The getter and setter for the query param state.
+ */
+function useQueryArg( key, initialValue ) {
```
Comment (tofumatt) [inline assets/js/hooks/useQueryArg.js:34]: I'd be inclined to add a third argument for dependency injection here, rather than the global mocking done in the test below. Something like `function useQueryArg( key, initialValue, _global = global ) { ` or something. What do you think?

File: assets/js/hooks/useQueryArg.test.js [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+		oldHistory = global.history;
+		delete global.location;
+		delete global.history;
+		global.location = {
+			href: 'http://example.com/path?page=demo',
+		};
+		global.history = {
+			replaceState: historyReplaceStateMock,
+		};
+	} );
```
Comment (tofumatt) [inline assets/js/hooks/useQueryArg.test.js:41]: As mentioned above—it might be a bit tidier to allow for the `global` variable in the hook to be injected for testing purposes rather than doing this global mocking dance.

File: assets/js/hooks/useQueryArg.test.js [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+
+	it( 'should push a new state to history when setQuery is called', () => {
+		const { result: { current: [ , setQuery ] } } = renderHook( ( ) => useQueryArg( 'page' ) );
+
+		actHook( () => {
+			setQuery( 'prod' );
+		} );
+
+		const updatedURL = 'http://example.com/path?page=prod';
+		expect( historyReplaceStateMock ).toHaveBeenCalledWith( null, '', updatedURL );
```
Comment (tofumatt) [inline assets/js/hooks/useQueryArg.test.js:66]: Hmm, I noticed the nested `const` object declaration in the function above and find it hard to read, but this is an even denser line. I think it's hard to follow `const { result: { current: [ , setQuery ] } }` especially, but the whole line is using so much shorthand that it reduces readability to me. I think this, and other functions in this file, can be expanded to allow for better readability.

Since these are tests we don't need to be too concise or even worry about DRY—the main aim of the tests should be the utmost readability and understanding at a glance to discern attention, even if it means they're a bit longer as to be clearer 🙂 

```suggestion
		const { result } = renderHook( ( ) => useQueryArg( 'page' ) );
		const [ , setQuery ] = result.current;

		actHook( () => {
			setQuery( 'prod' );
		} );

		const updatedURL = 'http://example.com/path?page=prod';
		expect( historyReplaceStateMock ).toHaveBeenCalledWith( null, '', updatedURL );
```

**Review by tofumatt — CHANGES_REQUESTED** [review]
This looks really good to me. I think mostly small tweaks are needed here. Mostly around test readability and the `useCallback` hook usage.

Nice 👍🏻

File: assets/js/hooks/useQueryArg.js (hunk 29 above)
Comment (kostyalmm) [reply assets/js/hooks/useQueryArg.js:34]: This actually is a great idea. Thanks :)

File: assets/js/components/settings/settings-admin.js [hunk 32]
```diff
@@ -21,18 +21,50 @@
...
+			{ featureFlags.userInput.enabled && (
+				<Cell size={ 12 }>
+					<Layout>
+						{ isUserInputCompleted ? (
+							<div className="googlesitekit-module-page googlesitekit-settings-user-input">
+								<UserInputPreview goTo={ goTo } noFooter />
+							</div>
+						) : (
+							<UserInputSettings isDimissable={ false } />
+						) }
```
Comment (tofumatt) [inline assets/js/components/settings/settings-admin.js:64]: In the future I'd prefer we not use ternaries in JSX and use something like:

```suggestion
						{ isUserInputCompleted && (
							<div className="googlesitekit-module-page googlesitekit-settings-user-input">
								<UserInputPreview goTo={ goTo } noFooter />
							</div>
						) }
						{ ! isUserInputCompleted && <UserInputSettings isDimissable={ false } /> }
```

**Review by tofumatt — APPROVED** [review]
Looks great! Thanks for the updates. 🙂

**Review by aaemnnosttv — APPROVED** [review]
LGTM 👍

File: assets/js/components/user-input/UserInputPreview.js [hunk 33]
```diff
@@ -58,7 +57,11 @@ export default function UserInputPreview( { back, goTo } ) {
 		setIsNavigating( true );
 		const response = await saveUserInputSettings();
 		if ( ! response.error ) {
-			global.location.assign( addQueryArgs( dashboardURL, { notification: 'user_input_success' } ) );
+			const url = new URL( redirectURL || dashboardURL );
+			// Here we don't use `addQueryArgs` due to a bug with how it handles hashes
+			// See https://github.com/WordPress/gutenberg/issues/16655
+			url.searchParams.set( 'notification', 'user_input_success' );
```
Comment (felixarntz) [inline assets/js/components/user-input/UserInputPreview.js:63]: We don't need to add this query parameter for the alternative redirect, it should only apply to the default `dashboardURL`. That query parameter isn't used elsewhere.

That way we can simplify this to only `addQueryArgs` to the default URL like before (if no `redirectURL` is passed), and use `redirectURL` as is.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Awesome work here so far! Just a few pieces of feedback, one unrelated to this PR, but worth addressing while working on this.

I also think we might wanna add a heading to the new box (see https://github.com/google/site-kit-wp/issues/2041#issuecomment-744765020), but that's pending so don't worry about it for now.

**Comment by benbowler** [comment]
@aaemnnosttv @tofumatt @felixarntz Are you happy for me to pick up this ticket from here to get it through CR? I'm working related work that it blocked by this.

**Comment by benbowler** [comment]
Also, I got wires crossed and have re-implemented this ticket using `useHash` as part of my work on #2499 - checkout branch `enhancement/2041-settings-block-user-input-ui-incorrect-ticket` - it seems super clean to do it this way.

**Comment by eugene-manuilov** [comment]
@googlebot I consent.

**Review by felixarntz — APPROVED** [review]
Apologies for the delay on reviewing this one.

The only thing missing here were the title and description as required via https://github.com/google/site-kit-wp/issues/2041#issuecomment-745357833. I fixed that in f64fafbccc3040b508e8b491d7b2c4621471fd8e, and also fixed the UI a little bit by using the grid instead of manually adding paddings via CSS.

Good for merge now!
